File: Substring with Concatenation of All Words/Substring with Concatenation of All Words.java
public class Solution {
	public List<Integer> findSubstring(String S, String[] L) {
        List<Integer> res = new ArrayList<Integer>();
        if (0 == L.length || 0 == L[0].length()) return res;
        
		Map<String, Integer> mainMap = new HashMap<String, Integer>();
		for (String s : L) {
			if (true == mainMap.containsKey(s)) mainMap.put(s, mainMap.get(s) + 1);
			else mainMap.put(s, 1);
		}
		
		int num = L.length;
		int len = L[0].length();
		HashMap<String, Integer> map = new HashMap<String, Integer>();

		for (int i = 0; i <= S.length() - num * len; ++i) {
			map.clear();
			for (int j = 0; j < L.length; ++j) {
				String word = S.substring(i + len * j, i + len * j + len);
				if (false == mainMap.containsKey(word)) break;
				else if (false == map.containsKey(word)) {
					map.put(word, 1);
				} else {
					map.put(word, map.get(word) + 1);
				}
				if (map.get(word) > mainMap.get(word)) break;
				if (j == L.length - 1) res.add(i);
			}
		}
        return res;
    }
}

File: Substring with Concatenation of All Words/Substring with Concatenation of All Words.cpp
class Solution {
public:
    vector<int> findSubstring(string S, vector<string> &L) {
        vector<int> res;
        if (0 == S.size() || 0 == L.size()) return res;
        
        int num = L.size();
        int len = L[0].size();
        unordered_map<string, int> main_map;
        for (string s : L) {
            if (main_map.find(s) == main_map.end()) main_map[s] = 1;
            else main_map[s] = main_map[s] + 1;
        }
        
        unordered_map<string, int> map;
        int ss = S.size();
        for (int i = 0; i <= ss - len * num; ++i) {
            map.clear();
            for (int j = 0; j < num; ++j) {
                string s = S.substr(i +  j * len, len);
                auto main_it = main_map.find(s);
                if (main_it == main_map.end()) break;
                auto it = map.find(s);
                if (it == map.end()) map[s] = 1;
                else {
                    it->second += 1;
                    if (it->second > main_it->second) break;
                }
                if (j == num - 1) res.push_back(i);
            }
        }
        return res;
    }
};